packages/file_selector/file_selector_windows/windows/file_selector_plugin.cpp

```diff
@@ -0,0 +1,358 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+#include "file_selector_plugin.h"
+
+#include <flutter/flutter_view.h>
+#include <flutter/method_channel.h>
+#include <flutter/plugin_registrar_windows.h>
+#include <flutter/standard_method_codec.h>
+#include <shobjidl.h>
+#include <windows.h>
+
+#include <cassert>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "file_dialog_controller.h"
+#include "string_utils.h"
+
+namespace file_selector_windows {
+
+namespace {
+
+using flutter::EncodableList;
+using flutter::EncodableMap;
+using flutter::EncodableValue;
+
+// From method_channel_file_selector.dart
+const char kChannelName[] = "plugins.flutter.io/file_selector";
+
+const char kOpenFileMethod[] = "openFile";
+const char kGetSavePathMethod[] = "getSavePath";
+const char kGetDirectoryPathMethod[] = "getDirectoryPath";
+
+const char kAcceptedTypeGroupsKey[] = "acceptedTypeGroups";
+const char kConfirmButtonTextKey[] = "confirmButtonText";
+const char kInitialDirectoryKey[] = "initialDirectory";
+const char kMultipleKey[] = "multiple";
+const char kSuggestedNameKey[] = "suggestedName";
+
+// From x_type_group.dart
+// Only 'extensions' are supported by Windows for filtering.
+const char kTypeGroupLabelKey[] = "label";
+const char kTypeGroupExtensionsKey[] = "extensions";
+
+// Looks for |key| in |map|, returning the associated value if it is present, or
+// a nullptr if not.
+const EncodableValue *ValueOrNull(const EncodableMap &map, const char *key) {
+  auto it = map.find(EncodableValue(key));
+  if (it == map.end()) {
+    return nullptr;
+  }
+  return &(it->second);
+}
+
+// Returns the path for |shell_item| as a UTF-8 string, or an
+// empty string on failure.
+std::string GetPathForShellItem(IShellItem *shell_item) {
+  wchar_t *wide_path = nullptr;
+  if (!SUCCEEDED(shell_item->GetDisplayName(SIGDN_FILESYSPATH, &wide_path))) {
+    return "";
+  }
+  std::string path = Utf8FromUtf16(wide_path);
+  ::CoTaskMemFree(wide_path);
+  return path;
+}
+
+// Implementation of FileDialogControllerFactory that makes standard
+// FileDialogController instances.
+class DefaultFileDialogControllerFactory : public FileDialogControllerFactory {
+ public:
+  DefaultFileDialogControllerFactory() {}
+  virtual ~DefaultFileDialogControllerFactory() {}
+
+  // Disallow copy and assign.
+  DefaultFileDialogControllerFactory(
+      const DefaultFileDialogControllerFactory &) = delete;
+  DefaultFileDialogControllerFactory &operator=(
+      const DefaultFileDialogControllerFactory &) = delete;
+
+  std::unique_ptr<FileDialogController> CreateController(
+      IFileDialog *dialog) const override {
+    return std::make_unique<FileDialogController>(dialog);
```

Comment:
Consider adding an assert not null here.

Reply:
Done.